File: src/main/java/io/github/erp/service/AmortizationPostingReportService.java
package io.github.erp.service;

/*-
 * Erp System - Mark X No 2 (Jehoiada Series) Server ver 1.7.1
 * Copyright © 2021 - 2023 Edwin Njeru and the ERP System Contributors (mailnjeru@gmail.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import io.github.erp.service.dto.AmortizationPostingReportDTO;
import java.util.Optional;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Service Interface for managing {@link io.github.erp.domain.AmortizationPostingReport}.
 */
public interface AmortizationPostingReportService {
    /**
     * Save a amortizationPostingReport.
     *
     * @param amortizationPostingReportDTO the entity to save.
     * @return the persisted entity.
     */
    AmortizationPostingReportDTO save(AmortizationPostingReportDTO amortizationPostingReportDTO);

    /**
     * Partially updates a amortizationPostingReport.
     *
     * @param amortizationPostingReportDTO the entity to update partially.
     * @return the persisted entity.
     */
    Optional<AmortizationPostingReportDTO> partialUpdate(AmortizationPostingReportDTO amortizationPostingReportDTO);

    /**
     * Get all the amortizationPostingReports.
     *
     * @param pageable the pagination information.
     * @return the list of entities.
     */
    Page<AmortizationPostingReportDTO> findAll(Pageable pageable);

    /**
     * Get the "id" amortizationPostingReport.
     *
     * @param id the id of the entity.
     * @return the entity.
     */
    Optional<AmortizationPostingReportDTO> findOne(Long id);

    /**
     * Delete the "id" amortizationPostingReport.
     *
     * @param id the id of the entity.
     */
    void delete(Long id);

    /**
     * Search for the amortizationPostingReport corresponding to the query.
     *
     * @param query the query of the search.
     *
     * @param pageable the pagination information.
     * @return the list of entities.
     */
    Page<AmortizationPostingReportDTO> search(String query, Pageable pageable);
}

File: erp-client/src/main/webapp/app/erp/erp-prepayments/amortization-posting-report/service/amortization-posting-report.service.ts
///
/// Erp System - Mark X No 11 (Jehoiada Series) Client 1.7.9
/// Copyright © 2021 - 2024 Edwin Njeru (mailnjeru@gmail.com)
///
/// This program is free software: you can redistribute it and/or modify
/// it under the terms of the GNU General Public License as published by
/// the Free Software Foundation, either version 3 of the License, or
/// (at your option) any later version.
///
/// This program is distributed in the hope that it will be useful,
/// but WITHOUT ANY WARRANTY; without even the implied warranty of
/// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
/// GNU General Public License for more details.
///
/// You should have received a copy of the GNU General Public License
/// along with this program. If not, see <http://www.gnu.org/licenses/>.
///

import { Injectable } from '@angular/core';
import { HttpClient, HttpParams, HttpResponse } from '@angular/common/http';
import { Observable } from 'rxjs';

import { isPresent } from 'app/core/util/operators';
import { ApplicationConfigService } from 'app/core/config/application-config.service';
import { createRequestOption } from 'app/core/request/request-util';
import { SearchWithPagination } from 'app/core/request/request.model';
import { IAmortizationPostingReport, getAmortizationPostingReportIdentifier } from '../amortization-posting-report.model';

export type EntityResponseType = HttpResponse<IAmortizationPostingReport>;
export type EntityArrayResponseType = HttpResponse<IAmortizationPostingReport[]>;

@Injectable({ providedIn: 'root' })
export class AmortizationPostingReportService {
  protected resourceUrl = this.applicationConfigService.getEndpointFor('api/prepayments/amortization-posting-reports/reported');
  protected resourceSearchUrl = this.applicationConfigService.getEndpointFor('api/prepayments/_search/amortization-posting-reports');

  constructor(protected http: HttpClient, protected applicationConfigService: ApplicationConfigService) {}

  findByDate(reportDate: string): Observable<EntityResponseType> {
    let reportParams: HttpParams = new HttpParams();
    reportParams = reportParams.set('reportDate', reportDate);

    return this.http.get<IAmortizationPostingReport>(`${this.resourceUrl}`, { params: reportParams, observe: 'response' });
  }

  find(id: number): Observable<EntityResponseType> {
    return this.http.get<IAmortizationPostingReport>(`${this.resourceUrl}/${id}`, { observe: 'response' });
  }

  query(req?: any): Observable<EntityArrayResponseType> {
    const options = createRequestOption(req);
    return this.http.get<IAmortizationPostingReport[]>(this.resourceUrl, { params: options, observe: 'response' });
  }

  queryByDate(req?: any): Observable<EntityArrayResponseType> {
    const options = createRequestOption(req);
    return this.http.get<IAmortizationPostingReport[]>(this.resourceUrl, { params: options, observe: 'response' });
  }

  search(req: SearchWithPagination): Observable<EntityArrayResponseType> {
    const options = createRequestOption(req);
    return this.http.get<IAmortizationPostingReport[]>(this.resourceSearchUrl, { params: options, observe: 'response' });
  }

  addAmortizationPostingReportToCollectionIfMissing(
    amortizationPostingReportCollection: IAmortizationPostingReport[],
    ...amortizationPostingReportsToCheck: (IAmortizationPostingReport | null | undefined)[]
  ): IAmortizationPostingReport[] {
    const amortizationPostingReports: IAmortizationPostingReport[] = amortizationPostingReportsToCheck.filter(isPresent);
    if (amortizationPostingReports.length > 0) {
      const amortizationPostingReportCollectionIdentifiers = amortizationPostingReportCollection.map(
        amortizationPostingReportItem => getAmortizationPostingReportIdentifier(amortizationPostingReportItem)!
      );
      const amortizationPostingReportsToAdd = amortizationPostingReports.filter(amortizationPostingReportItem => {
        const amortizationPostingReportIdentifier = getAmortizationPostingReportIdentifier(amortizationPostingReportItem);
        if (
          amortizationPostingReportIdentifier == null ||
          amortizationPostingReportCollectionIdentifiers.includes(amortizationPostingReportIdentifier)
        ) {
          return false;
        }
        amortizationPostingReportCollectionIdentifiers.push(amortizationPostingReportIdentifier);
        return true;
      });
      return [...amortizationPostingReportsToAdd, ...amortizationPostingReportCollection];
    }
    return amortizationPostingReportCollection;
  }
}
